File: packages/@eep-dev/middleware/src/event-bus/in-memory.ts
import type { CloudEvent, EventBusAdapter } from "../core/request-handler.js";

export class InMemoryEventBusAdapter implements EventBusAdapter {
  private readonly events: CloudEvent[] = [];
  private readonly subscribers: Array<{ pattern: string; handler: (event: CloudEvent) => void }> = [];

  async publish(event: CloudEvent): Promise<void> {
    this.events.push(event);
    for (const subscriber of this.subscribers) {
      if (this.matches(subscriber.pattern, event.type)) {
        subscriber.handler(event);
      }
    }
  }

  async subscribe(pattern: string, handler: (event: CloudEvent) => void): Promise<void> {
    this.subscribers.push({ pattern, handler });
  }

  getPublishedEvents(): CloudEvent[] {
    return [...this.events];
  }

  private matches(pattern: string, eventType: string): boolean {
    if (pattern === "*") {
      return true;
    }
    if (!pattern.includes("*")) {
      return pattern === eventType;
    }
    const prefix = pattern.split("*")[0];
    return eventType.startsWith(prefix);
  }
}

File: packages/eep-middleware-python/eep_middleware/event_bus/in_memory.py
from __future__ import annotations

from typing import Callable

from eep_middleware.adapters import CloudEvent


class InMemoryEventBusAdapter:
    def __init__(self) -> None:
        self._events: list[CloudEvent] = []
        self._subscribers: list[tuple[str, Callable[[CloudEvent], None]]] = []

    async def publish(self, event: CloudEvent) -> None:
        self._events.append(event)
        for pattern, handler in self._subscribers:
            if self._matches(pattern, event.event_type):
                handler(event)

    async def subscribe(self, pattern: str, handler: Callable[[CloudEvent], None]) -> None:
        self._subscribers.append((pattern, handler))

    def get_published_events(self) -> list[CloudEvent]:
        return list(self._events)

    @staticmethod
    def _matches(pattern: str, event_type: str) -> bool:
        if pattern == "*":
            return True
        if "*" not in pattern:
            return pattern == event_type
        prefix = pattern.split("*")[0]
        return event_type.startswith(prefix)
